p2p/simulations/adapters/inproc.go

```diff
@@ -34,6 +34,11 @@ import (
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
+const (
+	socketReadBuffer  = 5000 * 1024
+	socketWriteBuffer = 5000 * 1024
+)
```

Comment:
How are these numbers chosen?

Reply:
Based on experiments with the current message size and number of messages between nodes. Depending on the tests we want to run as part of the test suite, we can amend them.

When running pss_test.go TestNetwork with:
```
go test -v ./swarm/pss -cpu 4 -msgs 8000 -nodecount 3 -addrsize 20 -adapter sock -run TestNetwork
```
it fails with
```
const (
	socketReadBuffer  = 1000 * 1024
	socketWriteBuffer = 1000 * 1024
)
```
and starts passing with values above:
```
const (
	socketReadBuffer  = 2000 * 1024
	socketWriteBuffer = 2000 * 1024
)
```
depending on the concurrency profile.

If we revert and introduce delays between sending messages, probably we will need smaller buffers, but I left it at 5000 for now. We cannot have much higher values, because the OS will throw an error.

Reply:
I understand. So this should be a function of one or more of the parameters, perhaps?

Reply:
Well `socket` adapter could be used for other tests. It is a function of the message size we sent over that socket, and the number of messages. I am not sure we can easily configure it from parameters, without introducing a lot of complexity.

Reply:
Yes, of course, this setting is not in the `swarm/pss` folder %) I would imagine discovering performance thresholds is one of the goals of tests such as these, so maybe there's something to be said for examining this complexity? And also making these buffer sizes customizable by the calling code?

Reply:
@nolash I will look into making this customisable. However for now I think it is more important to understand why the `pss_test.go` actually need an asynchronous channel (and bigger buffers for that matter). I am under the impression that the tests should also pass with the `sim` adapter, which is synchronous and has no buffer at all.

Reply:
Sounds perfectly reasonable to defer the customization of parameters.

I'm not entirely sure of the full implications of what you mean when you write:

> why the pss_test.go actually need an asynchronous channel

Do you mean the fact that the `send` in the `sendAsym` was wrapped in a `go func()`, and fails if you remove it?

Should we create a separate issue for this, where the problem is explained in more detail? As mentioned below, my opinion is that we should only be fixing these problems on the `pss` branch if they are exclusive to `pss`.

Reply:
@nolash I think the reason for the difference of behaviour in PSS when run with synchronous (sim) versus asynchronous (tcp, exec) adapter, is the `self.Overlay.EachConn` function. On asynchronous channels, it can always iterate over all connections, whereas on synchronous channels it would block on the first neighbour.

Moving the `pp.Send()` inside a goroutine inside `self.Overlay.EachConn` changes this - even if a single communication channel blocks, we still get to send the given message to all neighbours with both synchronous and asynchronous adapters.

Reply:
Changing the unix socket buffer size now becomes redundant, because synchronous channels work as well, if we send a message to all neighbours in `EachConn`.

Reply:
Right, so this problem then still persists in the p2p layer in other words? AFAIR the issue of synchronicity in p2p sends was attempted adressed several months ago. We should open an issue for this, because it should not be necessary to wrap pp.Send asynchronously.

Reply:
@nolash it is pretty clear now that the `p2p` package is reading and then handling requests/messages synchronously, and this is by design. Each protocol's `handle` method can decide whether to handle requests synchronously or not, but this is outside of the `p2p` package.

Because of this, it is expected that PSS is deadlocking on the `sim` adapter - if Peers are handling a `forward` operation in a given point in time, they wouldn't be listening for new messages, and would deadlock - the handle operation will block and the EachConn operation will block. This is expected behaviour only if the connection buffer is full, or doesn't exist (sock or sim adapters)

We have to either:
1. Get rid of the `sim` adapter, if we assume we must have a buffer on the connection channel.
2. Add multi-message handling to the PSS protocol (essentially handling messages in a separate goroutine).

cc @zelig

Reply:
Can we create an issue for this please? Then we can discuss there. Deadlock has caused problems in other simulations aswell, so we need a fix that works for the sim-framework as a whole.

Reply:
https://github.com/ethersphere/go-ethereum/issues/130 - does that make sense to you?

In the meantime, I will split this PR into smaller chunks, which once addressed should resolve the deadlock and have a reasonable test suite on PSS that is stable.

Reply:
Can we keep the async pss send calls for now? Then address the *real* deadlock fix in a different branch off `swarm-network-rewrite`?

How did you think to chunk it up?

Reply:
Sure, we can keep the async pss send calls (and remove the async call from EachConn).

In terms of splitting the PR:
1. `tcp` adapter
2. updated `sock` adapter (with configurable read/write buffer sizes)
3. individual `pss` locks
4. changes to pss_test.go (including the new snapshots)
5. changes to rpc/client.go (configurable client subscription buffer) - mentioning Felix on this one.

only 3 and 4 are PSS related, however all changes have an effect on the pss_test.go

Reply:
@nolash I've split the PR into smaller ones so that it is easier to review. I've also added an explanation on the deadlock situation at https://github.com/nonsense/go-ethereum/blob/pss-networktest-tcp-and-unix-buffer-size/swarm/pss/writeup.md

I will keep this PR around just for reference and a placeholder of past comments. We should reach an agreement on how to solve the test failures / deadlocks (last item in the writeup document).